tests/migrations/test_loader.py

```diff
@@ -48,6 +54,28 @@ def test_apply(self):
             set(),
         )
 
+    def test_has_table_cached(self):
+        """
+        Test that the has_table() method caches a positive result and does not
+        continually query for the existence of the migrations table.
+        """
+        recorder = MigrationRecorder(connection)
+        # Ensure that the recorder table does not exist
+        with connection.schema_editor() as editor:
+            editor.delete_model(recorder.Migration)
+
+        with self.assertNumQueries(2):
+            self.assertEqual(recorder.has_table(), False)
+            self.assertEqual(recorder.has_table(), False)
```

Comment:
This is not properly isolated, if any of these assertions fail than dozens of test crashing with `no such table: django_migrations`.

Reply:
Hmm, I switched it to be a `TransactionTestCase` which should roll back the deletion of the `django_migrations` table?

But you're right, maybe it's simpler to do what you suggested in https://github.com/django/django/pull/16892#discussion_r1205317875

Reply:
You cannot rollback DDL operations on many databases.

Reply:
We could add:
```python
try:
    ...
finally:
    recorder.ensure_schema()
```

Reply:
Good point! I've changed the test to call `has_table()` twice and assert that only a single SQL query is executed instead.

Reply:
> We could add:

I can do that if you prefer, but in retrospect my initial test was overkill. I think the current test is simpler and tests the same stuff, but let me know if not.

Reply:
> > We could add:
> 
> I can do that if you prefer, but in retrospect my initial test was overkill. I think the current test is simpler and tests the same stuff, but let me know if not.

Fine for me :+1: